internal/kubernetes/operator/dbusers/dbusers.go

```diff
@@ -0,0 +1,128 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dbusers
+
+import (
+	"fmt"
+
+	"github.com/mongodb/mongodb-atlas-cli/internal/kubernetes/operator/secrets"
+	"github.com/mongodb/mongodb-atlas-cli/internal/store"
+	atlasV1 "github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1/common"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1/status"
+	"go.mongodb.org/atlas/mongodbatlas"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const PasswordField = "password"
+
+func BuildDBUsers(provider store.AtlasOperatorDBUsersStore, projectID, projectName, targetNamespace string, includeSecrets bool) ([]*atlasV1.AtlasDatabaseUser, []*corev1.Secret, error) {
+	users, err := provider.DatabaseUsers(projectID, &mongodbatlas.ListOptions{})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(users) == 0 {
+		return nil, nil, nil
+	}
+
+	var result []*atlasV1.AtlasDatabaseUser
+	var relatedSecrets []*corev1.Secret
+
+	for i := range users {
+		user := &users[i]
+		labels := convertUserLabels(user)
+		roles := convertUserRoles(user)
+		scopes := convertUserScopes(user)
+
+		secret := buildUserSecret(projectName, user, targetNamespace, includeSecrets)
+		relatedSecrets = append(relatedSecrets, secret)
+
+		result = append(result, &atlasV1.AtlasDatabaseUser{
+			TypeMeta: v1.TypeMeta{
+				Kind:       "AtlasDatabaseUser",
+				APIVersion: "atlas.mongodb.com/v1",
+			},
+			ObjectMeta: v1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-%s", projectName, user.Username),
+				Namespace: targetNamespace,
+			},
+			Spec: atlasV1.AtlasDatabaseUserSpec{
+				Project: common.ResourceRefNamespaced{
+					Name:      projectName,
+					Namespace: targetNamespace,
+				},
+				DatabaseName:    user.DatabaseName,
+				DeleteAfterDate: user.DeleteAfterDate,
+				Labels:          labels,
+				Roles:           roles,
+				Scopes:          scopes,
+				PasswordSecret: &common.ResourceRef{
+					Name: secret.Name,
+				},
+				Username: user.Username,
+				X509Type: user.X509Type,
+			},
+			Status: status.AtlasDatabaseUserStatus{},
+		})
+	}
+
+	return result, relatedSecrets, nil
+}
+
+func buildUserSecret(projectName string, user *mongodbatlas.DatabaseUser, targetNamespace string, includeSecrets bool) *corev1.Secret {
+	secret := secrets.NewAtlasSecret(fmt.Sprintf("%s-%s", projectName, user.Username), targetNamespace, map[string][]byte{
+		PasswordField: []byte(""),
+	})
+	if includeSecrets {
+		secret.Data[PasswordField] = []byte(user.Password)
+	}
+	return secret
+}
+
+func convertUserScopes(user *mongodbatlas.DatabaseUser) []atlasV1.ScopeSpec {
+	var result []atlasV1.ScopeSpec
+	for _, scope := range user.Scopes {
+		result = append(result, atlasV1.ScopeSpec{
+			Name: scope.Name,
+			Type: atlasV1.ScopeType(scope.Type),
+		})
+	}
```

Comment:
FYI if you know the size of `user.Scopes` preallocating `result` will save memory and will be faster

Reply:
I see this pattern in multiple loops which could be improved to use pre-allocation

Reply:
Fixed here and in a number of other locations